jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/testimpl/WebSocketHandler.java

```diff
@@ -0,0 +1,457 @@
+    {
+        Handshaker handshaker = getHandshaker(baseRequest);
+        if (LOG.isDebugEnabled())
+            LOG.debug("handle {} handshaker={}",baseRequest,handshaker);
+        if (handshaker!=null && handshaker.upgradeRequest(baseRequest, request, response))
+            return;
+        super.handle(target,baseRequest,request,response);
+    }
+
+
+    protected Handshaker getHandshaker(Request baseRequest)
+    {
+        // TODO this can eventually be made pluggable
+        HttpField version = baseRequest.getHttpFields().getField(HttpHeader.SEC_WEBSOCKET_VERSION);
+        if (version !=null && version.getIntValue()==RFC6455Handshaker.VERSION)
+            return RFC6455_Handshaker;
+        return null;
+    }
+
+
+    private final static class RFC6455Handshaker implements Handshaker
+    {
+
+        final static int VERSION = 13;
+
+        public boolean upgradeRequest(Request baseRequest, HttpServletRequest request, HttpServletResponse response) throws IOException
+        {
+            if (!HttpMethod.GET.is(request.getMethod()))
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("not upgraded method!=GET {}",baseRequest);
+                return false;
+            }
+
+            if (!HttpVersion.HTTP_1_1.equals(baseRequest.getHttpVersion()))
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("not upgraded version!=1.1 {}",baseRequest);
+                return false;
+            }
+
+            boolean upgrade = false;
+            QuotedCSV connectionCSVs = null;
+            String key = null;
+            QuotedCSV extensionCSVs = null;
+            String subprotocol = null;
+
+            for (HttpField field : baseRequest.getHttpFields())
+            {
+                if (field.getHeader()!=null)
+                {
+                    switch(field.getHeader())
+                    {
+                        case UPGRADE:
+                            if (!"websocket".equalsIgnoreCase(field.getValue()))
+                                return false;
+                            upgrade = true;
+                            break;
+
+                        case CONNECTION:
+                            if (connectionCSVs==null)
+                                connectionCSVs = new QuotedCSV();
+                            connectionCSVs.addValue(field.getValue().toLowerCase());
+                            break;
+
+                        case SEC_WEBSOCKET_KEY:
+                            key = field.getValue();
+                            break;
+
+                        case SEC_WEBSOCKET_EXTENSIONS:
+                            if (extensionCSVs==null)
+                                extensionCSVs = new QuotedCSV();
+                            extensionCSVs.addValue(field.getValue());
```

Comment:
Broken

Reply:
As above, I'd prefer to update/extend the utility class rather than duplicate.